### PR TITLE
[WIP] Http driver fixes

### PR DESCRIPTION
**Pull request by niranjanraoh**
- Porting the HTTP driver for M66
- Added HTTP file download, file open, read, close functions - so that we can read the received data chunk by chunk

### CODE REVIEW

File: firmware/ubirch/m66/m66_core.c [hunk 1]
```diff
@@ -46,13 +46,13 @@ void BOARD_CELL_UART_IRQ_HANDLER(void) {
     // it may be necessary to create a critical section here, but
     // right now it didn't hurt us to not disable interrupts
 
-    // __disable_irq();
```
Comment (thinkberg) [inline firmware/ubirch/m66/m66_core.c]: revert this please, it seems not necessary

Comment (thinkberg) [inline firmware/ubirch/m66/m66_core.c]: try to revert in a neutral way, so it does not show up here as a change

File: firmware/ubirch/m66/m66_http.c [hunk 2]
```diff
@@ -63,6 +65,111 @@ int modem_http(http_method_t method, size_t *res_size, uint32_t timeout) {
   return status;
 }
 
+int modem_http_file_dl(const char *file_name, uint32_t timeout)
+{
```
Comment (thinkberg) [inline firmware/ubirch/m66/m66_http.c]: move the `{` to the previous line end (same below)

File: firmware/ubirch/m66/m66_http.c [hunk 3]
```diff
@@ -63,6 +65,111 @@ int modem_http(http_method_t method, size_t *res_size, uint32_t timeout) {
...
+  timer_set_timeout(timeout * 1000);
+
+  int dl_size, content_len, dl_error_code = 0;
+
+  modem_send("AT+QHTTPDL=\"%s\"", file_name);
+
+  if (!modem_expect_OK(5 * 1000))
+  {
+    PRINTF("Failed to download file\r\n");
```
Comment (thinkberg) [inline firmware/ubirch/m66/m66_http.c]: use the CSTDEBUG() macros, not PRINTF here

File: firmware/ubirch/m66/m66_http.c [hunk 4]
```diff
@@ -63,6 +65,111 @@ int modem_http(http_method_t method, size_t *res_size, uint32_t timeout) {
...
+  return file_handle;
+}
+
+// TODO: add seek and position function, to get to the point where we need to start reading from
+size_t modem_http_file_read(char *read_buffer, int file_handle, size_t len, uint32_t timeout)
+{
+  timer_set_timeout(timeout * 1000);
+
+  char expect_connect[11] = {0};
+  snprintf(expect_connect, 11, "CONNECT %d", len);
```
Comment (thinkberg) [inline firmware/ubirch/m66/m66_http.c]: this seems unnecessary: use `modem_expect_scan` to parse the number below

File: firmware/ubirch/m66/m66_http.c [hunk 5]
```diff
@@ -63,6 +65,111 @@ int modem_http(http_method_t method, size_t *res_size, uint32_t timeout) {
...
+
+  modem_send("AT+QFREAD=%d,%d", file_handle, 100);
+
+  if (!modem_expect(expect_connect, uTimer_Remaining))
```
Comment (thinkberg) [inline firmware/ubirch/m66/m66_http.c]: shouldn't this be `if(!modem_expect_scan("CONNECT %d", &expected_len, ...)`
then check if expected_len == len

File: firmware/ubirch/m66/m66_http.h [hunk 6]
```diff
@@ -53,6 +53,37 @@ int modem_http_prepare(const char *url, uint32_t timeout);
  */
 int modem_http(http_method_t op, size_t *res_size, uint32_t timeout);
 
+
```
Comment (niranjanraoh) [inline firmware/ubirch/m66/m66_http.h]: i still have to document it, this was jus for reference for future documentation

**Comment by niranjanraoh** [comment]
RAM:file_name is not fixed yet

File: firmware/ubirch/m66/m66_http.c [hunk 7]
```diff
@@ -24,9 +24,13 @@
 #include <stddef.h>
 #include <stdint.h>
 #include <ubirch/timer.h>
+#include <stdio.h>
 #include "ubirch/modem.h"
 #include "m66_debug.h"
 
+
+char file_name[] = "RAM:file.txt";
```
Comment (thinkberg) [inline firmware/ubirch/m66/m66_http.c]: change this to something like file.bin as we are probably not downloading any text files, keeps confusion low.

Comment (niranjanraoh) [inline firmware/ubirch/m66/m66_http.c]: fixed

File: firmware/ubirch/m66/m66_http.c [hunk 8]
```diff
@@ -83,26 +194,67 @@ size_t modem_http_read(uint8_t *buffer, uint32_t start, size_t size, uint32_t ti
...
 
-  return modem_http(HTTP_GET, res_size, uTimer_Remaining);
+  modem_send("AT+QHTTPGET=%d", timer_timeout_remaining() / 1000);
+  if (!modem_expect_OK(uTimer_Remaining)) return 0;
+
+  dl_size = modem_http_file_dl(file_name, uTimer_Remaining);
+
+  if (!dl_size) return 0;
+
+  return dl_size;
```
Comment (thinkberg) [inline firmware/ubirch/m66/m66_http.c]: I don't think you should change the API return value here, the response size should be stored in the resize check that please.

**Comment by thinkberg** [comment]
Replaced by https://github.com/ubirch/ubirch-board-firmware/pull/11

**Pull request closed without merging**